apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/SessionPoolTests.cs

```diff
@@ -0,0 +1,443 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
+using Google.Api.Gax.Grpc;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Google.Protobuf;
+using Grpc.Core;
+using Moq;
+using Xunit;
+using Xunit.Abstractions;
+// ReSharper disable AccessToDisposedClosure
+
+namespace Google.Cloud.Spanner.Data.Tests
+{
+    public class SessionPoolTests
+    {
+        private static readonly DatabaseName s_defaultName =
+            DatabaseName.Parse("projects/project1/instances/instance1/databases/database1");
+
+        private ConcurrentDictionary<Mock<SpannerClient>, List<Session>> _createdSessions =
+            new ConcurrentDictionary<Mock<SpannerClient>, List<Session>>();
+
+        private ConcurrentStack<Transaction> _transactions = new ConcurrentStack<Transaction>();
+
+        public SessionPoolTests(ITestOutputHelper outputHelper)
+        {
+            //Uncomment these lines to debug a specific test.
+//            TestLogger.TestOutputHelper = outputHelper;
+//            TestLogger.Install();
+//            Logger.LogLevel = V1.Logging.LogLevel.Debug;
+        }
+
+        //Creates a client that will log created sessions in the _createdSession dictionary.
+        private Mock<SpannerClient> CreateMockClient(DatabaseName dbName = null)
+        {
+
+            var mockClient = new Mock<SpannerClient>();
+            if (dbName == null)
+            {
+                dbName = s_defaultName;
+            }
+            _createdSessions.TryAdd(mockClient, new List<Session>());
```

Comment:
its actually meant to track all sessions across all clients and done in a central place so that it doesnt need to be set up locally in each test.
At some point I'll add some handling to add clients not found as that was the initial intent.